src/code/z_play.c

```diff
@@ -1,81 +1,1575 @@
+        alpha = R_MOTION_BLUR_ALPHA;
+
+        if (sMotionBlurStatus == MOTION_BLUR_OFF) {
+            sMotionBlurStatus = MOTION_BLUR_SETUP;
+        }
+    } else {
+        alpha = 0;
+        sMotionBlurStatus = MOTION_BLUR_OFF;
+    }
+
+    if (sMotionBlurStatus != MOTION_BLUR_OFF) {
+        OPEN_DISPS(gfxCtx);
+
+        gfxHead = POLY_OPA_DISP;
+        gfx = Graph_GfxPlusOne(gfxHead);
+
+        gSPDisplayList(OVERLAY_DISP++, gfx);
+
+        this->pauseBgPreRender.fbuf = gfxCtx->curFrameBuffer;
+        this->pauseBgPreRender.fbufSave = this->unk_18E64;
+
+        if (sMotionBlurStatus == MOTION_BLUR_PROCESS) {
+            func_80170AE0(&this->pauseBgPreRender, &gfx, alpha);
+        } else {
+            sMotionBlurStatus = MOTION_BLUR_PROCESS;
+        }
+
+        func_801705B4(&this->pauseBgPreRender, &gfx);
+
+        gSPEndDisplayList(gfx++);
+
+        Graph_BranchDlist(gfxHead, gfx);
+
+        POLY_OPA_DISP = gfx;
+
+        CLOSE_DISPS(gfxCtx);
+    }
+}
+
+void Play_InitMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+    R_MOTION_BLUR_PRIORITY_ENABLED = false;
+    sMotionBlurStatus = MOTION_BLUR_OFF;
+}
+
+void Play_DestroyMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+    R_MOTION_BLUR_PRIORITY_ENABLED = false;
+    sMotionBlurStatus = MOTION_BLUR_OFF;
+}
+
+void Play_SetMotionBlurAlpha(u32 alpha) {
+    R_MOTION_BLUR_ALPHA = alpha;
+}
+
+void Play_EnableMotionBlur(u32 alpha) {
+    R_MOTION_BLUR_ALPHA = alpha;
+    R_MOTION_BLUR_ENABLED = true;
+}
+
+void Play_DisableMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+}
+
+/**
+ * Converts an RGBA16 buffer to an Intensity Image
+ *
+ * @param[out] destI destination buffer
+ * @param[in] srcRgba16 source buffer
+ * @param[in] rgba16Width width of a full row fo the RGBA16
+ * @param[in] pixelLeft X coordinate of the top-left RGBA16 pixel to start with
+ * @param[in] pixelTop Y coordinate of the top-left RGBA16 pixel to start with
+ * @param[in] pixelRight X coordinate of the bottom-right RGBA16 pixel to end with
+ * @param[in] pixelBottom Y coordinate of the bottom-right RGBA16 pixel to end with
+ * @param[in] bitDepth bit depth for the intensity image
+ */
+void Play_ConvertRgba16ToIntensityImage(void* destI, u16* srcRgba16, s32 rgba16Width, s32 pixelLeft, s32 pixelTop,
+                                        s32 pixelRight, s32 pixelBottom, s32 bitDepth) {
+    s32 i;
+    s32 j;
+    u32 pixel;
+    u32 r;
+    u32 g;
+    u32 b;
+
+    switch (bitDepth) {
+        case 4: {
+            u8* destI4 = (u8*)destI;
+            u32 upper;
+            u32 lower;
+
+            for (i = pixelTop; i <= pixelBottom; i++) {
+                for (j = pixelLeft; j <= pixelRight; j += 2) {
+                    pixel = srcRgba16[i * rgba16Width + j];
+                    r = RGBA16_GET_R(pixel);
+                    g = RGBA16_GET_G(pixel);
+                    b = RGBA16_GET_B(pixel);
+                    upper = ((r * 2 + g * 4 + b) * 0xF) / 217;
```

Comment:
217 is `0x1F * (2+4+1)`, it's for normalization

Reply:
Nice, makes sense. I wonder if this whole expression deserves a macro. 

Also, does it make sense to think of this as `(r << 1) + (g << 2) + (b << 0)`? Although I don't fully follow this conversion.

So in full:
`#define COLOR_UPPER(r, g, b) ( (((r << 1) + (g << 2) + (b << 0)) * 0xF) / (0x1F * (1 <<1 + (1 << 2) + (1 << 0))) )`?
except with a better macro name and double checking brackets? Maybe even a separate macro for the normalization constant? Or maybe it's all too much. Just throwing out ideas.

Reply:
why use shifts? the 2, 4, 1 are just weights
for example look at https://en.wikipedia.org/wiki/YUV#Conversion_to/from_RGB
Luminance is defined in that page as `Y = 0.299 * r + 0.587 * g + 0.114 * b`
which are close-ish to oot's 2,4,1 (after normalization)

This isn't news to me either, I've already seen this appear in oot's vismono
https://github.com/zeldaret/oot/blob/7ecafcfe7d357ff1cbcf49ce4d87a9efedae552e/src/code/z_vismono.c#L16

The reason for the calculation to not be (r+g+b)/3 btw, is the human eye sees green much better than the other colors (and red better than blue I guess (idk how the coefficients from wikipedia were derived))

Reply:
I think the macro idea sounds good. And as Dragorn pointed out it should probably not use shifts

Reply:
Seems good. I copied the macros from vismono into play (they could possibly just be combined and put in the color header, but I'm not 100% sure they should be correlated in that way). I've also added a `MIX` macro, which I like, but would like others opinions.